sdks/python/apache_beam/utils/annotations.py

```diff
@@ -0,0 +1,94 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import warnings
+from functools import partial
+from functools import wraps
+
+# produces only the first occurrence of matching warnings,
+# regardless of location per line of execution. Since the number of lines
+# of execution depends on the concrete runner, the number of warnings
+# produced will vary depending on the runner.
+warnings.simplefilter("once")
+
+
+def annotate(label, since, current):
+  """Decorates a function with a deprecated or experimental annotation.
+
+  The 'deprecated' annotation requires a 'since" parameter to specify
+  what version deprecated it.
+  Both 'deprecated' and 'experimental' annotations can specify the
+  current recommended version to use by means of a 'current' parameter.
+
+  The following example illustrates how to annotate coexisting versions of the
+  same function 'multiply'.
+  def multiply(arg1, arg2):
+    print arg1, '*', arg2, '=',
+    return arg1*arg2
+
+  # This annotation marks 'old_multiply' as deprecated since 'v.1' and suggests
+  # using 'multiply' instead.
+  @deprecated(since='v.1', current='multiply')
```

Comment:
Probably most of this comment and examples should be in a module level comment since users will not directly use the method annotate().